xwiki-platform-core/xwiki-platform-notifications/xwiki-platform-notifications-api/src/main/java/org/xwiki/notifications/internal/ScopeNotificationFilter.java

```diff
@@ -66,18 +66,64 @@
     @Override
     public boolean filterEvent(Event event, DocumentReference user, NotificationFormat format)
     {
+        return this.filterEventByFilterType(event, user, format, NotificationPreferenceScopeFilterType.EXCLUSIVE)
+                || this.filterEventByFilterType(event, user, format, NotificationPreferenceScopeFilterType.INCLUSIVE);
+    }
+
+    @Override
+    public String queryFilterOR(DocumentReference user, NotificationFormat format, String type)
+    {
+        return this.generateQueryString(user, format, type, NotificationPreferenceScopeFilterType.INCLUSIVE);
+    }
+
+    @Override
+    public String queryFilterAND(DocumentReference user, NotificationFormat format, String type)
+    {
+        return this.generateQueryString(user, format, type, NotificationPreferenceScopeFilterType.EXCLUSIVE);
+    }
+
+    @Override
+    public Map<String, Object> queryFilterParams(DocumentReference user, NotificationFormat format,
+            List<String> enabledEventTypes)
+    {
+        Map<String, Object> params =
+                this.generateQueryFilterParams(user, format, enabledEventTypes,
+                        NotificationPreferenceScopeFilterType.INCLUSIVE);
+        params.putAll(this.generateQueryFilterParams(user, format, enabledEventTypes,
+                NotificationPreferenceScopeFilterType.EXCLUSIVE));
+
+        return params;
+    }
+
+    /**
+     * Just as {@link #filterEvent(Event, DocumentReference, NotificationFormat)}, use the given user, the event, the
+     * format of the wanted notification and the type of filter we want to apply (see
+     * {@link NotificationPreferenceScopeFilterType}.
+     *
+     * @since 9.7RC1
+     */
+    private boolean filterEventByFilterType(Event event, DocumentReference user, NotificationFormat format,
+            NotificationPreferenceScopeFilterType scopeFilterType)
+    {
         // Indicate if a restriction exist concerning this type of event
         boolean hasRestriction = false;
         // Indicate if a restriction matches the document of the event
         boolean matchRestriction = false;
 
         try {
-            for (NotificationPreferenceScope scope : modelBridge.getNotificationPreferenceScopes(user, format)) {
+            for (NotificationPreferenceScope scope : modelBridge.getNotificationPreferenceScopes(user, format,
+                    scopeFilterType)) {
                 if (scope.getEventType().equals(event.getType())) {
                     hasRestriction = true;
```

Comment:
I think hasRestriction should be true only with NotificationPreferenceScopeFilterType.INCLUSIVE